src/components/datagrid/data_grid_types.ts

```diff
@@ -47,6 +50,29 @@ export interface EuiDataGridToolbarProps {
   displaySelector: ReactNode;
   columnSelector: ReactNode;
   columnSorting: ReactNode;
+  renderCustomToolbar?: (props: EuiDataGridCustomToolbarOptions) =>
+    | {
+        // to replace only controls inside the toolbar, use `left` and `right`
+        left?: ReactNode;
+        right?: ReactNode;
+      }
+    | ReactElement; // to replace the whole toolbar
+}
+
+/**
+ * Props which are available for a custom toolbar rendering
+ */
+export interface EuiDataGridCustomToolbarOptions {
```

Comment:
Rightly or wrongly, EUI uses `props` for a lot of the args/params we pass back from render callbacks - let's keep that naming for consistency

```suggestion
export interface EuiDataGridCustomToolbarProps {
```

Reply:
@cee-chen Updated via https://github.com/elastic/eui/pull/7190/commits/bbf2d184cba32b2dff36d512a4087bb18ac6add5